catkin_ws/src/state_estimation/src/state_aggregator.py

```diff
@@ -7,65 +7,76 @@
 from rosgraph_msgs.msg import Clock
 
 DEG_PER_RAD = 180 / np.pi
 
+def swap_main_sensor_message(axis_index, sensor_name):
+    global last_warning_message_time
+
+    if rospy.get_time() - last_warning_message_time[axis_index] > sensor_swap_warning_interval:
+            last_warning_message_time[axis_index] = rospy.get_time()
+            rospy.logwarn("Using {} for {}.".format(sensor_name, axis_names[axis_index]))
+
+
 def update_state(msg):
     if update_state_on_clock:
         global last_clock_msg_s
         global last_clock_msg_ns
         if last_clock_msg_s == msg.clock.secs and last_clock_msg_ns == msg.clock.nsecs: return
         last_clock_msg_s = msg.clock.secs
         last_clock_msg_ns = msg.clock.nsecs
-        
-    pub_dvl_sensor_status.publish(dvl.isActive())
-    pub_imu_sensor_status.publish(imu.isActive())
-    pub_depth_sensor_status.publish(depth_sensor.isActive())
-    x = None
-    y = None
-    z = None
+
+    position = [None, None, None]
     quaternion = None
     global last_error_message_time
 
-    for sensor in sensor_priorities["x"]:
-        if sensor.isActive():
-            x = sensor.x
-            break
-    for sensor in sensor_priorities["y"]:
-        if sensor.isActive():
-            y = sensor.y
-            break
-    for sensor in sensor_priorities["z"]:
-        if sensor.isActive():
-            z = sensor.z
-            break
-    for sensor in sensor_priorities["orientation"]:
-        if sensor.isActive():
-            quaternion = Quaternion(x = sensor.q_nwu_auv.x, y = sensor.q_nwu_auv.y, z = sensor.q_nwu_auv.z, w = sensor.q_nwu_auv.w)
-            euler_dvlref_dvl = transformations.euler_from_quaternion([sensor.q_nwu_auv.x, sensor.q_nwu_auv.y, sensor.q_nwu_auv.z, sensor.q_nwu_auv.w])
+    dvl_reading = dvl.get_reading()
+    imu_reading = imu.get_reading()
+    imu_front_camera_reading = imu_front_camera.get_reading()
+    depth_sensor_reading = depth_sensor.get_reading()
+
+    for axis_index in range(len(position)):
+        for sensor_index, sensor in enumerate(sensor_priorities[axis_names[axis_index]]):
+            if sensor.get_is_active():
+                position[axis_index] = sensor.get_reading()[axis_names[axis_index]]
+                if sensor_index != 0:
+                    swap_main_sensor_message(axis_index, sensor.get_sensor_name()) 
+                break
+
+    for sensor_index, sensor in enumerate(sensor_priorities["orientation"]):
+        if sensor.get_is_active():
+            sensor_reading = sensor.get_reading()
+            reading_quaternion = sensor_reading["quaternion"]
+            reading_angular_velocity = sensor_reading["angular_velocity"] 
+            
+            quaternion = Quaternion(x = reading_quaternion.x, y = reading_quaternion.y, z = reading_quaternion.z, w = reading_quaternion.w)
+            euler_dvlref_dvl = transformations.euler_from_quaternion([reading_quaternion.x, reading_quaternion.y, reading_quaternion.z, reading_quaternion.w])
```

Comment:
I didn't add orientation to the previous for loop because it uses 2 data readings instead of just one, so I think it's just better to leave it as it is